deploy/docker-ephemeral/build/Makefile

```diff
@@ -0,0 +1,217 @@
+STRETCHNAMES  := dynamodb_local
+# the names of the docker images that we're building that are based on debian.
+DEBNAMES      :=  $(JESSIENAMES) $(STRETCHNAMES)
+
+# the arches that we build for alpine.
+ALPINEARCHES  := amd64 386 arm32v6
+# images we build that are based on alpine.
+ALPINENAMES   := elasticsearch java_maven_node_python localstack minio
+
+# dependencies between docker images. <first_image>-<image_needed_to_build_first_image>
+PREBUILDS     := airdock_rvm-airdock_base airdock_fakesqs-airdock_rvm localstack-java_maven_node_python
+
+# manifest files don't work for these when they are finding the image they are based on.
+# by adding the name of the docker image here, we use the image:tag-<arch> format, instead of <arch>/image:tag.
+NOMANIFEST    := airdock_rvm airdock_fakesqs localstack
+
+# convert from debian architecture string to docker architecture string. 
+dockerarch=$(patsubst i%,%,$(patsubst armel,arm32v5,$(patsubst armhf,arm32v7,$(patsubst arm64,arm64v8,$(1)))))
+
+# the local architecture, in debian format. (i386, amd64, armel, armhf, arm64, ..)
+LOCALDEBARCH  := $(shell dpkg --print-architecture)
+# the local architecture, in docker format. (386, amd64, arm32v5, arm32v7, arm64v8, ...)
+LOCALARCH     ?= $(call dockerarch,$(LOCALDEBARCH))
+
+ifeq ($(LOCALARCH),)
+  $(error LOCALARCH is empty, you may need to supply it.)
+endif
+
+# FIXME: make this a section that depends on LOCALARCH, so we can allow these images to be built on native arm32.
+# note that qemu's x86_64 support is not strong enough to cross-build most things on i386.
+# these targets won't build on the system emulators for these arches. working with the qemu team to fix. they think it might be https://bugs.launchpad.net/qemu/+bug/1813398 .
+BADARCHSIM    := localstack-arm32v6 java_maven_node_python-arm32v6
+
+# set the targets, depending on the distro base specified. this is so that the debian images are built for all of the debian arches, and the alpine images for it's arches.
+ifeq ($(DIST),DEBIAN)
+  ARCHES=$(DEBARCHES)
+  NAMES=$(DEBNAMES)
+endif
+ifeq ($(DIST),ALPINE)
+  ARCHES=$(ALPINEARCHES)
+  NAMES=$(ALPINENAMES)
+endif
+
+# turn on experimental features in docker.
+export DOCKER_CLI_EXPERIMENTAL=enabled
+
+# allow for us to (ab)use $$* in dependencies of rules.
+.SECONDEXPANSION:
+
+# disable make's default builtin rules, to make debugging output cleaner.
+MAKEFLAGS += --no-builtin-rules
+
+# empty out the default suffix list, to make debugging output cleaner.
+.SUFFIXES:
+
+# too much haskell. returns first or second from <fst>-<snd>, respectively.
+fst=$(word 1, $(subst -, ,$(1)))
+snd=$(word 2, $(subst -, ,$(1)))
+
+# filter the list of archetectures, removing archetectures that we know do not work for a given docker image.
+goodarches=$(filter-out $(call snd,$(foreach arch,$(ARCHES),$(filter $(1)-$(arch),$(BADARCHSIM)))),$(ARCHES))
+
+.PHONY: manifest
+manifest: $(foreach name,$(NAMES),manifest-push-$(name))
```

Comment:
Do you have a good Makefile guide you can suggest? I understand the basics, but I'd love a resource where I can skim through advanced features; or is this just using shell expansion?